pkg/restore/restore_new_test.go

```diff
@@ -1662,6 +1664,425 @@ func assertRestoredItems(t *testing.T, h *harness, want []*test.APIResource) {
+// map and returns the corresponding volumeID if it exists, or an error otherwise.
+func (vs *volumeSnapshotter) CreateVolumeFromSnapshot(snapshotID, volumeType, volumeAZ string, iops *int64) (volumeID string, err error) {
+	volumeID, ok := vs.snapshotVolumes[snapshotID]
+	if !ok {
+		return "", errors.New("snapshot not found")
+	}
+
+	return volumeID, nil
+}
+
+// SetVolumeID sets the persistent volume's spec.awsElasticBlockStore.volumeID field
+// with the provided volumeID.
+func (*volumeSnapshotter) SetVolumeID(pv runtime.Unstructured, volumeID string) (runtime.Unstructured, error) {
+	unstructured.SetNestedField(pv.UnstructuredContent(), volumeID, "spec", "awsElasticBlockStore", "volumeID")
+	return pv, nil
+}
+
+// GetVolumeID panics because it's not expected to be used for restores.
+func (*volumeSnapshotter) GetVolumeID(pv runtime.Unstructured) (string, error) {
+	panic("GetVolumeID should not be used for restores")
+}
+
+// CreateSnapshot panics because it's not expected to be used for restores.
+func (*volumeSnapshotter) CreateSnapshot(volumeID, volumeAZ string, tags map[string]string) (snapshotID string, err error) {
+	panic("CreateSnapshot should not be used for restores")
+}
+
+// GetVolumeInfo panics because it's not expected to be used for restores.
+func (*volumeSnapshotter) GetVolumeInfo(volumeID, volumeAZ string) (string, *int64, error) {
+	panic("GetVolumeInfo should not be used for restores")
+}
+
+// DeleteSnapshot panics because it's not expected to be used for restores.
+func (*volumeSnapshotter) DeleteSnapshot(snapshotID string) error {
+	panic("DeleteSnapshot should not be used for backups")
+}
+
+// TestRestorePersistentVolumes runs restores for persistent volumes and verifies that
+// they are restored as expected, including restoring volumes from snapshots when expected.
+// Verification is done by looking at the contents of the API and the metadata/spec/status of
+// the items in the API.
+func TestRestorePersistentVolumes(t *testing.T) {
+	withReclaimPolicy := func(policy corev1api.PersistentVolumeReclaimPolicy) func(*corev1api.PersistentVolume) {
+		return func(pv *corev1api.PersistentVolume) {
+			pv.Spec.PersistentVolumeReclaimPolicy = policy
+		}
+	}
+	withClaimRef := func(ns, name string) func(*corev1api.PersistentVolume) {
+		return func(pv *corev1api.PersistentVolume) {
+			pv.Spec.ClaimRef = &corev1api.ObjectReference{
+				Namespace: ns,
+				Name:      name,
+			}
+		}
+	}
+	withVolumeName := func(volumeName string) func(*corev1api.PersistentVolume) {
+		return func(pv *corev1api.PersistentVolume) {
+			pv.Spec.AWSElasticBlockStore = &corev1api.AWSElasticBlockStoreVolumeSource{
+				VolumeID: volumeName,
+			}
+		}
+	}
+	withLabels := func(labels ...string) func(*corev1api.PersistentVolume) {
+		return func(pv *corev1api.PersistentVolume) {
+			test.WithLabels(labels...)(pv)
+		}
+	}
+
+	newPV := func(name string, opts ...func(*corev1api.PersistentVolume)) *corev1api.PersistentVolume {
```

Comment:
is there a difference between these helpers and the ones in pkg/test/resources?

Reply:
The difference is that these take functional options that can specifically customize PV spec/status, rather than any generic API object's metadata. These should probably move into `pkg/test` and be combined with the other ones in some way, my brain's just too tired to figure out a reasonable way to do it at the moment :)

Reply:
Ah I see, yeah I can't see a better way to do this either right now. Maybe worth adding a comment to revisit this?

Reply:
I'm looking at this again this morning. One option I'm considering is to turn these into functional options similar to what's already in `pkg/test/resources.go`, that take an arg of type `metav1.Object`, but to type-assert that object to a PV and to panic if it fails. That keeps the signature for using these simple but provides immediate feedback if they're used incorrectly. And since it's test code, I don't have any issue with just panicking. An example:

```go
func WithReclaimPolicy(policy corev1.PersistentVolumeReclaimPolicy) func(obj metav1.Object) {
	return func(obj metav1.Object) {
		pv, ok := obj.(*corev1.PersistentVolume)
		if !ok {
			panic("WithReclaimPolicy is only valid for persistent volumes")
		}

		pv.Spec.PersistentVolumeReclaimPolicy = policy
	}
}
```

Thoughts?

Reply:
Went ahead and made this change, see latest commit.

Reply:
Nice, yeah I was wondering if it would be possible to use a type assertion here. This is much nicer now!